### PR TITLE
Support true forced rebuilds for vendored dependencies, refactor types

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

Fixes ANE-1737:

- Refactors the types for archive uploads so that archive upload and first party license scans share the same types.
- Refactors the now shared type to stop passing booleans around everywhere.
- Instructs Core to force a rescan of an archive if configured.

## Acceptance criteria

- The repro case in https://fossa.atlassian.net/browse/ANE-1737 works.
- Nothing else breaks.

## Testing plan

I manually tested the repro case in https://fossa.atlassian.net/browse/ANE-1737 and validated it works.

I am relying on tests and Haskell's strong typing to validate nothing breaks.

## Risks

I don't think it's very risky.

## Metrics

None.

## References

Resolves https://fossa.atlassian.net/browse/ANE-1737.

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this PR added docs, I added links as appropriate to the user manual's ToC in `docs/README.ms` and gave consideration to how discoverable or not my documentation is.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json` AND I have updated example files used by `fossa init` command. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).
- [x] If I made changes to a subcommand's options, I updated `docs/references/subcommands/<subcommand>.md`.

### CODE REVIEW

File: src/App/Fossa/LicenseScanner.hs [hunk 1]
```diff
@@ -351,11 +352,11 @@ licenseScanSourceUnit ::
   ) =>
   VendoredDependencyScanMode ->
   Maybe LicenseScanPathFilters ->
-  FullFileUploads ->
+  FileUpload ->
   Path Abs Dir ->
   NonEmpty VendoredDependency ->
   m (NonEmpty Locator)
-licenseScanSourceUnit vendoredDependencyScanMode licenseScanPathFilters fullFileUploads baseDir vendoredDeps = do
+licenseScanSourceUnit mode licenseScanPathFilters upload baseDir vendoredDeps = do
```
Comment (spatten) [inline src/App/Fossa/LicenseScanner.hs:359]: I think renaming this to `skipMode` would make future-me have to think less about what `mode` means

**Review by spatten — APPROVED** [review]
This is a great change! The types are much better, and it just makes sense to support forcing rebuilds on archive uploads.

I have a few minor nits about variable names, and one more important comment about making sure that it's impossible to run an archive upload with anything other than `FileUploadFullContent`.

File: src/App/Fossa/ArchiveUploader.hs [hunk 2]
```diff
@@ -76,10 +77,12 @@ archiveUploadSourceUnit ::
   , Has Logger sig m
   , Has FossaApiClient sig m
   ) =>
+  FileUpload ->
+  DependencyRebuild ->
   Path Abs Dir ->
   NonEmpty VendoredDependency ->
   m (NonEmpty Locator)
-archiveUploadSourceUnit baseDir vendoredDeps = do
+archiveUploadSourceUnit upload rebuild baseDir vendoredDeps = do
```
[the remark this answers is not in the document]
Comment (jssblck) [reply src/App/Fossa/ArchiveUploader.hs]: _excellent_ catch, thank you! reformatted- please let me know if that looks closer to the semantics you'd expect us to express. the short version is that we pass the rebuild flag all the way down.

File: src/App/Fossa/Analyze/Upload.hs [hunk 3]
```diff
@@ -167,11 +165,11 @@ uploadAnalysisWithFirstPartyLicensesToS3AndCore ::
   ProjectRevision ->
   ProjectMetadata ->
   NE.NonEmpty FullSourceUnit ->
-  FullFileUploads ->
+  FileUpload ->
   m UploadResponse
-uploadAnalysisWithFirstPartyLicensesToS3AndCore revision metadata mergedUnits fullFileUploads = do
+uploadAnalysisWithFirstPartyLicensesToS3AndCore revision metadata mergedUnits uploadKind = do
```
[the remark this answers is not in the document]
Comment (jssblck) [reply src/App/Fossa/Analyze/Upload.hs:170]: Done!

File: src/App/Fossa/LicenseScanner.hs (hunk 1 above)
Comment (jssblck) [reply src/App/Fossa/LicenseScanner.hs]: Done!

File: src/App/Fossa/ArchiveUploader.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (spatten) [reply src/App/Fossa/ArchiveUploader.hs]: That looks perfect. 🚢

**Pull request merged**